akka-remote/src/main/scala/akka/remote/artery/compress/OutboundActorRefCompressionTable.scala

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.remote.artery.compress
+
+import akka.actor.{ Address, ActorRef, ActorSystem }
+import akka.event.Logging
+
+import scala.annotation.tailrec
+
+final class OutboundActorRefCompressionTable(system: ActorSystem, remoteAddress: Address)
+  extends OutboundCompressionTable[ActorRef](system, remoteAddress) {
+
+  preAllocate(
+    system.deadLetters → 0
+  )
+
+  //  if (system.toString.contains("systemB"))
+  //    system.log.error(new Throwable, "new OutboundActorRefCompressionTable   = " + this.hashCode())
+
+  def preAllocate(allocations: (ActorRef, Int)*): Unit =
+    allocations foreach { case (ref, id) ⇒ register(ref, id) }
+}
+
+/**
+ * Base class for all outgoing compression.
+ * Encapsulates the compressedId registration and lookup.
+ *
+ * Not thread safe.
+ */
+class OutboundCompressionTable[T](system: ActorSystem, remoteAddress: Address) {
+  import OutboundCompressionTable._
+
+  private val settings = CompressionSettings(system)
+
+  private val log = system.log
+
+  // TODO can we specialize this? (tuning due here)
+  @volatile private[this] var backing = Map.empty[T, Int] // TODO could use unsafe to swap the map instead of volatile
+
+  // mapping guarding
+  private[this] var compressionIdsAllocated = -1
+  private[this] var aheadAllocatedCompressionIds = Set.empty[Int]
+
+  def register(value: T, id: Int): Unit = {
+    backing.get(value) match {
+      case None if isNextCompressionId(id) ⇒
+        log.debug("Outbound: Registering new compression from [{}] to [{}].", value, id) // TODO should be debug
+        addFastForwardCompressionIdsAllocatedCounter()
+        backing = backing.updated(value, id)
+
+        if (settings.debug) log.debug("Outgoing: Updated compression table state: \n{}", toDebugString) // TODO debug  
+
+      case None ⇒
+        // TODO could be wrong? since we can not guarantee alocations come in sequence?
```

Comment:
This TODO will be irrelevant very soon (entire compression advertised in one go)